builtin/providers/vsphere/resource_vsphere_virtual_disk.go

```diff
@@ -0,0 +1,213 @@
+						errors = append(errors, fmt.Errorf(
+							"only 'thin' and 'eagerZeroedThick' are supported values for 'type'"))
+					}
+					return
+				},
+			},
+
+			"datacenter": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"datastore": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceVSphereVirtualDiskCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*govmomi.Client)
+
+	vDisk := virtualDisk{
+		size: d.Get("size").(int),
+	}
+
+	if v, ok := d.GetOk("vmdk_path"); ok {
+		vDisk.vmdkPath = v.(string)
+	}
+
+	if v, ok := d.GetOk("type"); ok {
+		vDisk.initType = v.(string)
+	}
+
+	if v, ok := d.GetOk("datacenter"); ok {
+		vDisk.datacenter = v.(string)
+	}
+
+	if v, ok := d.GetOk("datastore"); ok {
+		vDisk.datastore = v.(string)
+	}
+
+	diskPath := fmt.Sprintf("[%v] %v", vDisk.datastore, vDisk.vmdkPath)
+
+	err := createHardDisk(client, vDisk.size, diskPath, vDisk.initType, vDisk.datacenter)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(diskPath)
+
+	return resourceVSphereVirtualDiskRead(d, meta)
+}
+
+func resourceVSphereVirtualDiskRead(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] Reading virtual disk.")
+	client := meta.(*govmomi.Client)
+
+	vDisk := virtualDisk{
+		size: d.Get("size").(int),
+	}
+
+	if v, ok := d.GetOk("vmdk_path"); ok {
+		vDisk.vmdkPath = v.(string)
+	}
+
+	if v, ok := d.GetOk("type"); ok {
+		vDisk.initType = v.(string)
+	}
+
+	if v, ok := d.GetOk("datacenter"); ok {
+		vDisk.datacenter = v.(string)
+	}
+
+	if v, ok := d.GetOk("datastore"); ok {
+		vDisk.datastore = v.(string)
+	}
+
+	dc, err := getDatacenter(client, d.Get("datacenter").(string))
+	if err != nil {
+		return err
+	}
+
+	finder := find.NewFinder(client.Client, true)
+	finder = finder.SetDatacenter(dc)
+
+	ds, err := finder.Datastore(context.TODO(), d.Get("datastore").(string))
+	if err != nil {
+		return err
+	}
+
+	fileInfo, err := ds.Stat(context.TODO(), vDisk.vmdkPath)
+	if err != nil {
+		log.Printf("[DEBUG] resourceVSphereVirtualDiskRead - stat failed on: %v", vDisk.vmdkPath)
+		d.SetId("")
+		return err
+	} else {
+		log.Printf("[FINDME] READ - fileinfo: %v", fileInfo.GetFileInfo())
+		// Todo collect and write size, type info
+	}
+	return nil
+
+}
+
+func resourceVSphereVirtualDiskDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*govmomi.Client)
+
+	vDisk := virtualDisk{}
+
+	if v, ok := d.GetOk("vmdk_path"); ok {
+		vDisk.vmdkPath = v.(string)
+	}
+	if v, ok := d.GetOk("datastore"); ok {
+		vDisk.datastore = v.(string)
+	}
+
+	dc, err := getDatacenter(client, d.Get("datacenter").(string))
+	if err != nil {
+		return err
+	}
+	diskPath := fmt.Sprintf("[%v] %v", vDisk.datastore, vDisk.vmdkPath)
+
+	virtualDiskManager := object.NewVirtualDiskManager(client.Client)
+
+	virtualDiskManager.DeleteVirtualDisk(context.TODO(), diskPath, dc)
```

Comment:
Does the API return an error on this call if it fails? For instance we try to delete a disk that is not there.